### ISSUE TITLE
change how bare `{}` is handled so `find -exec {}` isn't broken

### ISSUE THREAD

**Issue by hrldcpr** [fish-shell/fish-shell#1109]
"reopening" #95 

i think it's bad that simple, standard find commands fail with a perplexing error message in fish, due to its expansion of empty curly braces to empty string. e.g.:

```
> find . -name '*.txt' -exec cat {} +
find: -exec: no terminating ";" or "+"
```

as discussed in #95 the fix is to quote '{}' but this is kind of hard to figure out

**Comment by faho** [comment]
Duplicate of #434, though both titles are just special cases of a general problem: Brace expansion is parsed in an unintuitive (for bash/zsh users) manner.

**Comment by stephane-chazelas** [comment]
Note that brace expansion was introduced by csh. Csh (back in the end of the 70s) explicitly made a documented special case for `{}` alone to be left alone (though `{}foo` would expand to `foo`).

All other shells that implemented brace expansion thereafter (tcsh, ksh, zsh, bash, yash at least)  copied that behaviour (`{}` left alone, not the `{}foo` -> `foo` part (and `{foo}` -> `foo`)).

I can't imagine anyone relying on `{}` expanding to an empty argument. I'd expect people would use `''` or `""` for that instead, so I don't expect aligning with other shells would break anything (the `eval` case seems a bit contrived).

<sup>`rc` is another shell where `{}` needs to be quoted, but that's not for brace expansion there (`{` and `}` are used in the syntax of the language).</sup>

**Comment by krader1961** [comment]
See also issue #3002. It's always dangerous to change long established behavior. But in this particular case I think it is warranted. The `find ... -exec some_command {}` use case is particularly compelling because people don't expect the shell to eat the braces and doing so can have catastrophic consequences. Can anyone provide a good example of how changing fish to not elide a bare, free standing, `{}` would break fish scripts? If not I recommend we change fish to behave like bash in this regard.

**Comment by jbarlow83** [comment]
I agree that is a mistake in fish's design that should be corrected.

Tab completion is great, but does not help with people converting scripts or copy-pasting command lines.

Here's an example of a safe bash command that turns into `rm -rf /*` in fish (I gave this example earlier up).

```fish
find . -type d -exec rm -f {}/* \;   # do not do this in fish
```

**Comment by stephane-chazelas** [comment]
> Here's an example of a safe bash command that turns into rm -rf /* in fish (I gave this example earlier up).
> 
>     find . -type d -exec rm -f {}/* \;   # do not do this in fish

That one works the same in csh/tcsh (the original implementation of brace expansion) as in fish though. In csh, only `{}` on it's own is left alone.

Also note that `find -exec rm -f '{}/*' \;` would not work as `rm` doesn't  do globbing, and `find -exec cmd '{}word' \;` is not standard. (portably, you can only rely on `{}` on its own to be expanded).

**Comment by ridiculousfish** [comment]
It seems like we're all in favor of making {} expand to nothing. Assigning this to myself so I remember to fix it, but if someone would like to take it on, by all means please feel free.

**Comment by krader1961** [comment]
> It seems like we're all in favor of making {} expand to nothing.

Since that's the current behavior I assume you meant to say "making {} expand to itself."

**Comment by jbarlow83** [comment]
@ridiculousfish {} already expands to nothing; the suggestion is to make {} expand to {} since all other shells seem to that.

**Comment by krader1961** [comment]
Also, TBD is whether we should follow csh or bash: `echo {}foo` yields `foo` in csh and `{}foo` in bash.

**Comment by ridiculousfish** [comment]
Derp, yes, I meant expand to itself. Comment corrected. Thanks.

With that corrected, naively I would expect `echo {}foo` to expand to itself. Making `echo {}foo` expand to `foo` seems like a weird special case - is there any reason to prefer that?

**Comment by stephane-chazelas** [comment]
If you want to align with ksh/zsh/bash/yash, there's also the question of `echo {foo}`, and then of `echo {$var}` which at the moment `fish` recommends as the equivalent of `${var}` when not quoted.

Currently in `fish`, `{}` is a list of one empty element and `{$a}` is the list of elements of `$a` (With `set a`, `{$a}` is an empty list, with `set a ''`, that's a list of one empty element like `{}`). In that context, having `{}var` expand to `{}var`, when `set a; {$a}foo`  expands to nothing and `set a ''; {$a}foo` expands to `foo` may seem strange.

Note that there are some differences between all the ksh (ksh93, pdksh/mksh), bash, zsh and yash -o braceexpand implementations.

Generally for all those, `{...}` must contain at least one `,` to count as a brace expansion. I suppose the rationale is because you don't need brace expansion if there's only one value in them (`{foo}bar` is the same as `foobar`}.

Now, there's the question of the interaction with other expansions. Except maybe in zsh and yash (which behave more like `fish`), brace expansion looks more like an after-thought.

In bash, brace expansion is done very early which can cause some surprises like

```
$ bash -c 'a=a b=b; echo {$a,$b}-c'
a-c b-c
$ bash -c 'a=a b=b; echo {$a,$b}_c'

$
```

(That's interpreted as `echo $a_c $b_c`, you'd actually need:

```
bash -c 'a=a b=b; echo {${a},${b}}_c
```

or better:

```
bash -c 'a=a b=b; echo {"$a","$b"}_c
```

`bash` is also the only shell to give this result to this kind of construct:

```
$ bash -c 'echo {a,b,c}$((++x))'
a1 b2 c3
$ bash -c 'echo {a,b,c}$(readlink /proc/self)'
a27241 b27243 c27245
```

(those arithmetic and command substitutions evaluated several times)

Playing with the limits of the parser there can be interesting:

```
~$ bash -c 'echo $[{1+1,2*3,4/4}]'
2 6 1
~$ bash -c 'echo $(({1+1,2*3,4/4}))'
bash: {1+1,2*3,4/4}: syntax error: operand expected (error token is "{1+1,2*3,4/4}")
```

In `ksh`, that's the opposite problem:

```
$ ksh93 -c 'a=1,2; echo {$a,3}x'
1x 2x 3x
```

Quoting helps in pdksh/mksh, but not ksh93:

```
$ ksh93 -c 'a=1,2; echo {"$a",3}x'
1x 2x 3x
$ mksh -c 'a=1,2; echo {"$a",3}x'
1,2x 3x
```

In `mksh`, it can get worse though:

```
mksh -c 'a="1,2}{4"; echo {$a,3}x'
14x 13x 24x 23x
```


One thing I noticed is that all shells (including `fish` and `csh`) treat `{a,{1,2},b}` as `{a,1,2,b}`. However with both `fish` and `zsh`:

```
$ fish -c 'set v a b; echo {1,$v,2}x'
1x ax 2x 1x bx 2x
$ zsh -c 'v=(a b); echo {1,$^v,2}x'
1x ax 2x 1x bx 2x
```

Another thing `fish` and `zsh` have in common:

```
$ zsh -c 'echo ~ma{n,il}'
/var/cache/man /var/mail
$ fish -c 'echo ~ma{n,il}'
/var/cache/man /var/mail
```

(bash as well, but that's not surprising for bash)

**Comment by zanchey** [comment]
If we're doing a 3.0 (syntax-breaking) release, let's add this.

**Comment by krader1961** [comment]
> If we're doing a 3.0 (syntax-breaking) release, let's add this. 

Agreed, although there is the question of exactly which behavior we want. For example, [csh or bash](https://github.com/fish-shell/fish-shell/issues/1109#issuecomment-238056592).

**Comment by faho** [comment]
I don't see how csh's behavior is useful here.

We want to change the behavior for a literal `{}` because that is currently utterly useless - it always expands to nothing, so why use it in the first place? That also applies to `{}foo` expanding to `foo` - what use is the `{}` prefix there?

Its also consistent to have `{}` expand to itself (or "not expand" depending on how you look at it) in both cases - when standing alone, and when used as part of a word.

That means:

```fish
echo {} # prints {}
echo foo{}bar # prints foo{}bar
echo foo{a,{}}bar # prints fooabarr foo{}bar
echo foo{{}}bar # prints foo{}bar
```

In other words, something is only a brace expansion if it matches `{.+}`

**Comment by asymmetric** [comment]
So do I understand correctly that this change would have to wait for a major version to be released?

**Comment by faho** [comment]
Strictly speaking, yes. And... good news: The next fish version is planned to be 3.0.

**Comment by mqudsi** [comment]
I ran into an edge case yesterday where I actually used the bare {} syntax - I wanted to do something to files that shared a sequence number with files of a different name but the same sequence number, so I actually used this syntax:

```bash
rm file1_{(for n in (seq 20 100)
    if test -e file2_$n
        echo $n
    end
end)}
```

except I didn’t use `rm`. The real problem isn’t that `{}` is useless (even aside from the fact that using two for loops would have been more idiomatic and easier on the eyes) but that you almost never can use the _result_ of an empty expansion since the behavior of Unix commands generally changes when the number of arguments provided dwindles down to zero, ie the command above actually doesn’t do what I would want since passing zero arguments to rm is not the same as not executing rm. 

Basically, even when the intended expansion of braces is an empty set, they are still useless 😃
